Include max pool in fusion groups

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52614 Avoid dispatch overhead in call to mkldnn convolution
* **#52613 Include max pool in fusion groups**
* #52513 Extend subgraph utils to cover merging a node following a subgraph
* #52512 Remove unused subgraph vmap api
* #51924 Only functionalize ops which we want to include in mkldnn group
* #51923 Add a filter to remove mutation
* #51922 Dont error on 0-dim in convolution
* #51921 Add 0-dim support for binary mkldnn ops
* #51736 Handle mkldnn broadcasting in mkldnn fuser
* #51600 Add MKLDNN fuser
* #52786 [NNC] Make NNC sanitize input names
* #51484 Add support for linear in mkldnn fusion
* #51483 Move conv to mkldnn

Including MaxPool as part of the MKLDNN fusion group sped up resnet18 by ~20%, and was a win on other models I tested as well. I will post more complete benchmarks. 

As mentioned in the diff, in some cases MaxPool can be slower than aten - ideally we'd only include maxpool if it decreased the number of layout transformations that occur. That hasnt actually matttered for all of the torchvision models, I don't think its necessary for this PR.

Differential Revision: [D26696704](https://our.internmc.facebook.com/intern/diff/D26696704)

Commit:
Include max pool in fusion groups

[ghstack-poisoned]

Commit:
Update on "Include max pool in fusion groups"


Including MaxPool as part of the MKLDNN fusion group sped up resnet18 by ~20%, and was a win on other models I tested as well. I will post more complete benchmarks. 

As mentioned in the diff, in some cases MaxPool can be slower than aten - ideally we'd only include maxpool if it decreased the number of layout transformations that occur. That hasnt actually for all of the torchvision models, I don't think its necessary for this PR. 





[ghstack-poisoned]

Commit:
Update on "Include max pool in fusion groups"


Including MaxPool as part of the MKLDNN fusion group sped up resnet18 by ~20%, and was a win on other models I tested as well. I will post more complete benchmarks. 

As mentioned in the diff, in some cases MaxPool can be slower than aten - ideally we'd only include maxpool if it decreased the number of layout transformations that occur. That hasnt actually for all of the torchvision models, I don't think its necessary for this PR. 





[ghstack-poisoned]

Commit:
Update on "Include max pool in fusion groups"


Including MaxPool as part of the MKLDNN fusion group sped up resnet18 by ~20%, and was a win on other models I tested as well. I will post more complete benchmarks. 

As mentioned in the diff, in some cases MaxPool can be slower than aten - ideally we'd only include maxpool if it decreased the number of layout transformations that occur. That hasnt actually for all of the torchvision models, I don't think its necessary for this PR. 





[ghstack-poisoned]

Commit:
Update on "Include max pool in fusion groups"


Including MaxPool as part of the MKLDNN fusion group sped up resnet18 by ~20%, and was a win on other models I tested as well. I will post more complete benchmarks. 

As mentioned in the diff, in some cases MaxPool can be slower than aten - ideally we'd only include maxpool if it decreased the number of layout transformations that occur. That hasnt actually for all of the torchvision models, I don't think its necessary for this PR. 





[ghstack-poisoned]

Krovatkin commented on torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:
```diff
@@ -380,9 +380,17 @@ class MKLDNNSubgraphSlicer {
+    switch (n->kind()) {
+      case aten::relu:
+      case aten::sigmoid:
+      // TODO: max_pool on mkldnn can be slower than in eager. ideally, we'd
```
do we have data how often max_pool is slower than eager and by how much?

eellison replied:
i can regenerate it

Krovatkin reviewed (approved):
:shipit:

Commit:
Update on "Include max pool in fusion groups"


Including MaxPool as part of the MKLDNN fusion group sped up resnet18 by ~20%, and was a win on other models I tested as well. I will post more complete benchmarks. 

As mentioned in the diff, in some cases MaxPool can be slower than aten - ideally we'd only include maxpool if it decreased the number of layout transformations that occur. That hasnt actually matttered for all of the torchvision models, I don't think its necessary for this PR. 





[ghstack-poisoned]

Commit:
Update on "Include max pool in fusion groups"


Including MaxPool as part of the MKLDNN fusion group sped up resnet18 by ~20%, and was a win on other models I tested as well. I will post more complete benchmarks. 

As mentioned in the diff, in some cases MaxPool can be slower than aten - ideally we'd only include maxpool if it decreased the number of layout transformations that occur. That hasnt actually matttered for all of the torchvision models, I don't think its necessary for this PR. 





[ghstack-poisoned]

Commit:
Update on "Include max pool in fusion groups"


Including MaxPool as part of the MKLDNN fusion group sped up resnet18 by ~20%, and was a win on other models I tested as well. I will post more complete benchmarks. 

As mentioned in the diff, in some cases MaxPool can be slower than aten - ideally we'd only include maxpool if it decreased the number of layout transformations that occur. That hasnt actually matttered for all of the torchvision models, I don't think its necessary for this PR. 





[ghstack-poisoned]

Commit:
Update on "Include max pool in fusion groups"


Including MaxPool as part of the MKLDNN fusion group sped up resnet18 by ~20%, and was a win on other models I tested as well. I will post more complete benchmarks. 

As mentioned in the diff, in some cases MaxPool can be slower than aten - ideally we'd only include maxpool if it decreased the number of layout transformations that occur. That hasnt actually matttered for all of the torchvision models, I don't think its necessary for this PR.

Differential Revision: [D26696704](https://our.internmc.facebook.com/intern/diff/D26696704)

[ghstack-poisoned]
